File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int helper(int n, vector<int>& dp) {
        if (n == 0)
            return dp[n] = 1;
        if (n < 0)
            return 0;
        if (dp[n] != -1)
            return dp[n];
        int left = helper(n - 1, dp);
        int right = helper(n - 2, dp);
        return dp[n] = left + right;
    }
    int climbStairs(int n) {
        vector<int> dp(n + 1, -1);
        return helper(n, dp);
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {

    public int helper(int n,int[] dp){
        if(n<0)
            return 0;
        if(n==0)
            return 1;
        if(dp[n]!=-1)
            return dp[n];
        int twoSteps=helper(n-2,dp);
        int oneSteps=helper(n-1,dp);
        return dp[n]= twoSteps+oneSteps;
    }
    public int climbStairs(int n) {
        int[] dp=new int[n+1];
        Arrays.fill(dp,-1);
        return helper(n,dp);
    }
}